sys/net/application_layer/dhcpv6/client.c

```diff
@@ -0,0 +1,834 @@
+static event_t request = { .handler = _request };
+static event_t renew = { .handler = _renew };
+
+void dhcpv6_client_auto_init(void)
+{
+    thread_create(_thread_stack, DHCPV6_CLIENT_STACK_SIZE,
+                  DHCPV6_CLIENT_PRIORITY, THREAD_CREATE_STACKTEST,
+                  _thread, NULL, "dhcpv6-client");
+}
+
+void dhcpv6_client_init(event_queue_t *eq, uint16_t netif)
+{
+    assert(eq->waiter != NULL);
+    event_queue = eq;
+    local.netif = netif;
+    remote.netif = netif;
+}
+
+void dhcpv6_client_start(void)
+{
+    uint32_t delay = random_uint32_range(0, DHCPV6_SOL_MAX_DELAY * US_PER_SEC);
+
+    _generate_duid();
+    sock_udp_create(&sock, &local, NULL, 0);
+    timer.callback = _post_solicit_servers;
+    xtimer_set(&timer, delay);
+}
+
+void dhcpv6_client_req_ia_pd(uint16_t netif, uint8_t pfx_len)
+{
+    pfx_lease_t *lease = NULL;
+
+    for (unsigned i = 0; i < DHCPV6_CLIENT_PFX_LEASE_MAX; i++) {
+        if (pfx_leases[i].parent.ia_id.id == 0) {
+            lease = &pfx_leases[i];
+            lease->parent.ia_id.info.netif = netif;
+            lease->parent.ia_id.info.type = DHCPV6_OPT_IA_PD;
+            lease->pfx_len = pfx_len;
+        }
+    }
+}
+
+/* TODO make GNRC-independent */
+static void _configure_prefix(pfx_lease_t *lease, uint32_t valid,
+                              uint32_t pref)
+{
+    gnrc_netif_t *netif = gnrc_netif_get_by_pid(lease->parent.ia_id.info.netif);
+    eui64_t iid;
+    ipv6_addr_t addr;
+
+    assert(netif != NULL);
+    DEBUG("DHCPv6 client: (re-)configure prefix %s/%d\n",
+          ipv6_addr_to_str(addr_str, &lease->pfx, sizeof(addr_str)),
+          lease->pfx_len);
+    if (gnrc_netapi_get(netif->pid, NETOPT_IPV6_IID, 0, &iid,
+                        sizeof(eui64_t)) >= 0) {
+        ipv6_addr_set_aiid(&addr, iid.uint8);
+    }
+    else {
+        LOG_WARNING("DHCPv6 client: cannot get IID of netif %u\n", netif->pid);
+        return;
+    }
+    ipv6_addr_init_prefix(&addr, &lease->pfx, lease->pfx_len);
+    if (gnrc_netif_ipv6_addr_add(netif, &addr, lease->pfx_len, 0) > 0) {
+        /* update lifetime */
+        gnrc_ipv6_nib_pl_set(netif->pid, &lease->pfx, lease->pfx_len,
+                             valid, pref);
+#if defined(MODULE_GNRC_IPV6_NIB) && GNRC_IPV6_NIB_CONF_6LBR && \
+        GNRC_IPV6_NIB_CONF_MULTIHOP_P6C
+        gnrc_ipv6_nib_abr_add(&addr);
+#endif
+#ifdef MODULE_GNRC_RPL
+        gnrc_rpl_init(netif->pid);
+        gnrc_rpl_instance_t *inst = gnrc_rpl_instance_get(GNRC_RPL_DEFAULT_INSTANCE);
+        if (inst) {
+            gnrc_rpl_instance_remove(inst);
+        }
+        gnrc_rpl_root_init(GNRC_RPL_DEFAULT_INSTANCE, &addr, false, false);
+#endif
+    }
+}
+
+static void *_thread(void *args)
+{
+    (void)args;
+    event_queue_t event_queue;
+    /* TODO: make GNRC-independent */
+    gnrc_netif_t *border_interface = NULL;
+    gnrc_netif_t *wpan_interface = NULL;
+    gnrc_netif_t *netif = NULL;
+
+    while ((netif = gnrc_netif_iter(netif))) {
```

Comment:
After thinking about this a bit, I think this thread is very application specific to the border router, so I think I will just implement a default thread here and provide an alternative implementation in the border router example application when I port that.